llvm/include/llvm/Support/SYCLPropertySetIO.h

```diff
@@ -0,0 +1,268 @@
+
+  SYCLPropertyValue &operator=(SYCLPropertyValue &&P);
+
+  SYCLPropertyValue &operator=(const SYCLPropertyValue &P);
+
+  // Get property value as unsigned 32-bit integer
+  uint32_t asUint32() const {
+    assert((Ty == UINT32) && "must be UINT32 value");
+    return std::get<uint32_t>(Val);
+  }
+
+  // Get raw data size in bits.
+  SizeTy getByteArraySizeInBits() const {
+    assert((Ty == BYTE_ARRAY) && "must be BYTE_ARRAY value");
+    SizeTy Res = 0;
+
+    for (size_t I = 0; I < sizeof(SizeTy); ++I) {
+      auto ByteArrayVal = std::get<std::byte *>(Val);
+      Res |= (SizeTy)ByteArrayVal[I] << (8 * I);
+    }
+    return Res;
+  }
+
+  // Get byte array data size in bytes.
+  SizeTy getByteArraySize() const {
+    SizeTy SizeInBits = getByteArraySizeInBits();
+    constexpr unsigned int MASK = 0x7;
+    return ((SizeInBits + MASK) & ~MASK) / 8;
+  }
+
+  // Get byte array data size in bytes, including the leading bytes encoding the
+  // size.
+  SizeTy getRawByteArraySize() const {
+    return getByteArraySize() + sizeof(SizeTy);
+  }
+
+  // Get byte array data including the leading bytes encoding the size.
+  const std::byte *asRawByteArray() const {
+    assert((Ty == BYTE_ARRAY) && "must be BYTE_ARRAY value");
+    auto *ByteArrayVal = std::get<std::byte *>(Val);
+    return ByteArrayVal;
+  }
+
+  // Get byte array data excluding the leading bytes encoding the size.
+  const std::byte *asByteArray() const {
+    assert((Ty == BYTE_ARRAY) && "must be BYTE_ARRAY value");
+
+    auto ByteArrayVal = std::get<std::byte *>(Val);
+    return ByteArrayVal + sizeof(SizeTy);
+  }
+
+  bool isValid() const { return getType() != NONE; }
+
+  // Set property value when data type is UINT32_T
+  void set(uint32_t V) {
+    assert((Ty == UINT32) && "must be UINT32 value");
+    Val = V;
+  }
+
+  // Set property value when data type is BYTE_ARRAY
+  void set(std::byte *V, int DataSize) {
+    assert((Ty == BYTE_ARRAY) && "must be BYTE_ARRAY value");
+    size_t DataBitSize = DataSize * CHAR_BIT;
+    constexpr size_t SizeFieldSize = sizeof(SizeTy);
+    // Allocate space for size and data.
+    Val = new std::byte[SizeFieldSize + DataSize];
+
+    // Write the size into first bytes.
+    for (size_t I = 0; I < SizeFieldSize; ++I) {
+      auto ByteArrayVal = std::get<std::byte *>(Val);
+      ByteArrayVal[I] = (std::byte)DataBitSize;
+      DataBitSize >>= CHAR_BIT;
+    }
+    // Append data.
+    auto ByteArrayVal = std::get<std::byte *>(Val);
+    std::memcpy(ByteArrayVal + SizeFieldSize, V, DataSize);
+  }
+
+  Type getType() const { return Ty; }
+
+  SizeTy size() const {
+    switch (Ty) {
+    case UINT32:
+      return sizeof(uint32_t);
+    case BYTE_ARRAY:
+      return getRawByteArraySize();
+    default:
+      llvm_unreachable_internal("unsupported SYCL property type");
+    }
+  }
+
+private:
+  void copy(const SYCLPropertyValue &P);
+
+  Type Ty = NONE;
+  std::variant<uint32_t, std::byte *> Val;
```

Comment:
It's a little redundant to store the type and a variant (which itself stores the type). Should this just be a union?

Reply:
We initiallly used an 'union' and are trying to 'modernize' the code by using variants. I agree that it is redundant t store the type when using a variant.
Will try to resolve this soon.

Thanks